### ISSUE TITLE
Unexpected token in IMAP response: NIL

### ISSUE THREAD

**Issue by quentez** [jstedfast/MailKit#244]
I'm having an error with a Gmail inbox that returns the following content (it stops abruptly because it  crashes):

```
325 FETCH (X-GM-THRID 1507597005640876725 X-GM-LABELS (Palazzani Pennelli Rigamonti) UID 355 MODSEQ (112864) FLAGS (\Seen) ENVELOPE ("Fri, 24 Jul 2015 17:49:03 +0100" "Canarias OFFICINE" (("ligia moreau" NIL "ligiamoreau" "domain.es")) (("ligia moreau" NIL "ligiamoreau" "moreaurep.es")) (("ligia moreau" NIL "ligiamoreau" 
```

The error is:

```
MailKit.Net.Imap.ImapProtocolException: Unexpected token in IMAP response: NIL
Sep 08 15:29:08 WIN-0VVBLT697LQ FrontPumps:     at MailKit.Net.Imap.ImapUtils.ParseLabelsList(ImapEngine engine, CancellationToken cancellationToken)
Sep 08 15:29:08 WIN-0VVBLT697LQ FrontPumps:     at MailKit.Net.Imap.ImapFolder.FetchSummaryItems(ImapEngine engine, ImapCommand ic, Int32 index)
```

Any idea why this happens?

**Comment by jstedfast** [comment]
I don't think it's crashing on that line... it's probably crashing on one of the preceding lines.

**Comment by jstedfast** [comment]
My guess is that there's an `X-GM-LABELS NIL` or `X-GM-LABELS (label1 NIL label3)`-type scenario

**Comment by quentez** [comment]
Could that be it?

```
316 FETCH (X-GM-THRID 1507843069465059557 X-GM-LABELS ("\\Starred" NIL) UID 346 MODSEQ (112817)
```

That's the line just before the one I previously posted.

**Comment by jstedfast** [comment]
Yea, that is very likely the problem... I'm not sure what to do with a NIL label. Should I just ignore it?

**Comment by quentez** [comment]
Sounds appropriate... I'm not sure what it could mean to the client who set it in the first place.

**Comment by jstedfast** [comment]
I ended up including it in the list for completeness

**Comment by quentez** [comment]
Awesome! Thank you as always for the super-quick fix.

**Comment by bitdisaster** [comment]
I still this kind of exception a lot. Is that a known issue?

**Comment by jstedfast** [comment]
@bitdisaster all of the known cases have been fixed. What case are you hitting?

I'll likely need a protocol log so I know what token is `NIL` so I can figure out how to work around it. Most of these exceptions are caused by IMAP servers not adhering to the specifications (although in the original case for this bug report, it was simply that Google did not document their extension at all and so I was left to simply guess at the expected token types).

**Comment by bitdisaster** [comment]
In my case its mostly O365. How do I get you a log?

**Comment by jstedfast** [comment]
You can do something like this:

```csharp
var client = new ImapClient (new ProtocolLogger ("imap.log"));
```

**Comment by bitdisaster** [comment]
I have a hard time finding something in the log related to the error. Anything specific I should look for?
Here is what I'm doing:
 if (messageEmailPair.Key.HtmlBody != null && messageEmailPair.Value != null)
                            {
                                var bodyPart = (TextPart)await remoteFolder.GetBodyPartAsync(messageEmailPair.Key.UniqueId, messageEmailPair.Key.HtmlBody);
                                messageEmailPair.Value.HtmlBody = bodyPart.Text;
                            }
                            else if (messageEmailPair.Key.TextBody != null && messageEmailPair.Value != null)
                            {
                                var bodyPart = (TextPart)await remoteFolder.GetBodyPartAsync(messageEmailPair.Key.UniqueId, messageEmailPair.Key.TextBody);
                                messageEmailPair.Value.HtmlBody = bodyPart.Text;
                            }

messageEmailPair.Key is a MessageSummary I fetched before. It looks like I fail when I see emails with a calendar invite. Her an example of teh body parts where I see the exception.

{("text" "plain" ("charset" "utf-8") NIL NIL "base64" 0 NIL NIL NIL NIL 0)}
{("text" "calendar" ("charset" "utf-8" "method" "REPLY" "name" "meeting.ics") NIL NIL "base64" 1536 NIL NIL NIL NIL 20)}

Basically GetBodyPartAsync for {("text" "plain" ("charset" "utf-8") NIL NIL "base64" 0 NIL NIL NIL NIL 0)} fails.

**Comment by jstedfast** [comment]
The protocol error will be the last response from the server (not necessarily the last line because the code buffers 4K at a time from the socket).

In the IMAP log, you'll see MailKit send a command like this:

```
A0000027 UID FETCH <the unique id> (BODY.PEEK[1.2.MIME] BODY.PEEK[1.2])
```

And the response from the server will look something like:

```
* <message index> FETCH (UID <the unique id> BODY[1.2.MIME] {<size>}
<blurb containing the MIME headers> BODY[1.2] {<size>}
<blurb containing the content of the MIME part>
A0000027 FETCH OK success
```

**Comment by bitdisaster** [comment]
C: C00000074 UID FETCH 449 (BODY.PEEK[1.MIME] BODY.PEEK[1])
S: * 215 FETCH (BODY[1.MIME] {90}
S: Content-Transfer-Encoding: quoted-printable
S: Content-Type: text/plain; charset="utf-8"
S: 
S:  BODY[1] NIL UID 449)

**Comment by jstedfast** [comment]
Cool, thanks, that should be fixable.

**Comment by jstedfast** [comment]
Fixed.

**Comment by bitdisaster** [comment]
Awesome! You rock!
We have one more slightly different error. Something about a "atom" token. I will try to get a log for that one as well.

**Comment by jstedfast** [comment]
I've released MailKit 1.2.13 to NuGet with the fix for this bug.
